runtime/elem/builtins/helpers/GainFade.h

```diff
@@ -6,62 +6,116 @@
-            : currentGain(other.currentGain), targetGain(other.targetGain), step(other.step)
+        : currentGain(other.currentGain.load())
+        , targetGain(other.targetGain.load())
+        , step(other.step.load())
+        , inStep(other.inStep)
+        , outStep(other.outStep)
         {
         }
 
         void operator= (GainFade const& other)
         {
-            currentGain = other.currentGain;
-            targetGain = other.targetGain;
-            step = other.step;
+            currentGain.store(other.currentGain.load());
+            targetGain.store(other.targetGain.load());
+            step.store(other.step.load());
+            inStep = other.inStep;
+            outStep = other.outStep;
         }
 
         FloatType operator() (FloatType x) {
-            if (currentGain == targetGain)
-                return (currentGain * x);
+            auto const _currentGain = currentGain.load();
+            if (_currentGain == targetGain.load())
+                return (_currentGain * x);
 
-            auto y = x * currentGain;
-            currentGain = std::clamp(currentGain + step, FloatType(0), FloatType(1));
+            auto y = x * _currentGain;
+            currentGain.store(std::clamp(_currentGain + step.load(), FloatType(0), FloatType(1)));
 
             return y;
         }
 
-        void setTargetGain (FloatType g) {
-            targetGain = g;
+        void process (const FloatType* input, FloatType* output, int numSamples) {
+            auto const _currentGain = currentGain.load();
+            if (_currentGain == targetGain.load()) {
+                for (int i = 0; i < numSamples; ++i) {
+                    output[i] = input[i] * _currentGain;
+                }
+                return;
+            }
 
-            if (targetGain < currentGain) {
-                step = FloatType(-1) * std::abs(step);
-            } else {
-                step = std::abs(step);
+            auto const _step = step.load();
+            for (int i = 0; i < numSamples; ++i) {
+                output[i] = input[i] * std::clamp(_currentGain + _step * i, FloatType(0), FloatType(1));
             }
+            
+            currentGain.store(std::clamp(_currentGain + _step * numSamples, FloatType(0), FloatType(1)));
+        }
+
+        void setFadeInTimeMs(double sampleRate, double fadeInTimeMs) {
+            inStep = detail::millisecondsToStep(sampleRate, fadeInTimeMs);
+            updateCurrentStep();
+        }
+
+        void setFadeOutTimeMs(double sampleRate, double fadeOutTimeMs) {
+            outStep = FloatType(-1) * detail::millisecondsToStep(sampleRate, fadeOutTimeMs);
+            updateCurrentStep();
+        }
+
+        void fadeIn() {
+            targetGain.store(FloatType(1));
+            updateCurrentStep();
+        }
+
+        void fadeOut() {
+            targetGain.store(FloatType(0));
+            updateCurrentStep();
+        }
+
+        void setCurrentGain(FloatType gain) {
```

Comment:
Unused?

Reply:
we use it a couple places in `StereoSampleSeq`, e.g. https://github.com/splice/platform/blob/7bfd56361fcf97ebfd7a0dcd03b2a434c6b478a5/plugins/studio-one-integration/SpliceSampleSeq.h#L59

Reply:
Ah ty!